### PR TITLE
[aot] Device capability refactorization

### PR DESCRIPTION
**Pull request by PENGUINLIONG**
This PR refactorized `DeviceCapability` control logics in Taichi. `DeviceCapability` controls the compilation of kernels and the execution behavior on device. The current implementation collect capabilities from the device interface but it blocks extension of AOT code generation because it depends on the currently attached compute device whether to enable a set of feature during compilation. AOT users, however, demands an interface to control the availability of `DeviceCapability`.

This PR does several things and it seems large:
- Removed runtime capability form the list of caps, see code comments for details of design decisions;
- Encapsulate the capability settings in `Device` in `DeviceCapabilityConfig`;
- Replace any reference to `Device` with `DeviceCapabilityConfig` in codegens;
- Removed `TargetDevice` for in the SPIR-V codegen;
- Adapt backends (basically gfx backends) to this change.

Python and C-API users will be able to configure `DeviceCapability`s in future PRs.

### CODE REVIEW

File: python/taichi/examples/simulation/fractal.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import taichi as ti
 
-ti.init(arch=ti.gpu)
+ti.init(arch=ti.vulkan)
```
Comment (ailzhang) [inline python/taichi/examples/simulation/fractal.py:3]: accidental change?

File: taichi/rhi/vulkan/vulkan_device.cpp [hunk 2]
```diff
@@ -1329,6 +1329,7 @@ struct VulkanDevice::ThreadLocalStreams {
 VulkanDevice::VulkanDevice()
     : compute_streams_(std::make_unique<ThreadLocalStreams>()),
       graphics_streams_(std::make_unique<ThreadLocalStreams>()) {
+  caps_.set(DeviceCapability::spirv_version, 0x10000);
```
Comment (ailzhang) [inline taichi/rhi/vulkan/vulkan_device.cpp:1332]: Is this newly enforced btw?

**Review by ailzhang — APPROVED** [review]
Awesome!!

File: taichi/rhi/vulkan/vulkan_device.cpp (hunk 2 above)
Comment (PENGUINLIONG) [reply taichi/rhi/vulkan/vulkan_device.cpp:1332]: that's the minimum requirement. That capacity is required for the backend to work iirc

File: python/taichi/examples/simulation/fractal.py (hunk 1 above)
Comment (PENGUINLIONG) [reply python/taichi/examples/simulation/fractal.py:3]: If

**Comment by k-ye** [comment]
Awesome!

**Pull request merged**